### DISCUSSION TITLE
An example of implementing sql.Scanner

### DISCUSSION THREAD

**Discussion by Bravilogy** [pocketbase/pocketbase#3891] [Q&A] [answered]
Hello.

I have a custom struct:
```
type Fixture struct {
	models.BaseModel

	ID          uint         `json:"id" db:"id"`
	Slug        string       `json:"slug" db:"slug"`
	Points      int          `json:"points" db:"points"`
	Public      bool         `json:"public" db:"public"`

	Announcements AnnouncementDetails `json:"announcements" db:"announcements"`
        ...
}
```

And in my pocketbase schema, I have `announcements` defined as `json` field. From other discussions, I understand that I need to implement `sql.Scanner` for `AnnouncementDetails` to make this work. I was wondering if there's an example of how to do that? Even a basic one would help.

**Comment by dschissler** [comment]
@Bravilogy I regret wasting time with custom record structs.  There are rough edges and pitfalls without much benefit.  It gets nasty around hooks because it comes in as a record.  Until something changes about the situation then I recommend against it.  I wasted a lot of time.

**Reply by Bravilogy** [reply]
@dschissler is that for nested structs or just structs all together? I guess I could work off of just `record`, but I'm just thinking ahead of types and potential type casts all over the place..

**Reply by dschissler** [reply]
@Bravilogy You the normal `model.Record` that all all events make available.  I'm saying not to do what you're doing, but that's just my opinion.

**Comment by ganigeorgiev** [comment]
@Bravilogy At the moment, I'd advice against using custom models, unless you really know what you are doing. The custom models were initially created as a solution for the cases when you have a custom select query that merge, combine or aggregate data from multiple collections.

If you want to use custom model structs just for the type safety to access your collection data then you'll be better with `models.Record`. There will be major refactoring of the internals after the Presentator release (https://github.com/pocketbase/pocketbase/discussions/3271) and eventually we'll have autogenerated Go types based on the proposal from https://github.com/pocketbase/pocketbase/discussions/2697.

If you still want an example for the `driver.Valuer` and `sql.Scanner` interfaces you can check the `types.DateTime` implementation https://github.com/pocketbase/pocketbase/blob/985ab1e5b79a328b516decca3d1d7dc8edfcbba0/tools/types/datetime.go#L68-L104.

**Reply by ganigeorgiev** [reply, marked as answer]
I also forgot to mention that since `announcements` is a `json` field, in your case you can safely unmarshal its result in your custom `AnnouncementDetails` struct using the [`UnmarshalJSONField`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.19.4/models#Record.UnmarshalJSONField) method:

```go
announcements := AnnouncementDetails{}
record.UnmarshalJSONField("announcements", &announcements)
```
